fdbserver/ptxn/TLogPeekCursor.actor.cpp

```diff
@@ -185,4 +294,164 @@ bool ServerTeamPeekCursor::hasRemainingImpl() const {
+	cursorPtrs.emplace_back(std::move(pCursor));
+	auto iter = std::prev(std::end(cursorPtrs));
+	if ((*iter)->hasRemaining()) {
+		addCursorToCursorHeap(iter);
+	}
+	return iter;
+}
+
+MergedServerTeamPeekCursor::MergedServerTeamPeekCursor() : MergedPeekCursor() {}
+
+MergedPeekCursor::CursorContainer::iterator MergedServerTeamPeekCursor::addCursorImpl(
+    std::unique_ptr<PeekCursorBase>&& cursor) {
+
+	ASSERT(dynamic_cast<ServerTeamPeekCursor*>(cursor.get()) != nullptr);
+
+	auto iter = MergedPeekCursor::addCursorImpl(std::move(cursor));
+
+	const TeamID& teamID = dynamic_cast<ServerTeamPeekCursor*>((*iter).get())->getTeamID();
+	ASSERT(teamIDCursorMapper.find(teamID) == teamIDCursorMapper.end());
+	teamIDCursorMapper[teamID] = iter;
+
+	return iter;
+}
+
+std::unique_ptr<PeekCursorBase> MergedServerTeamPeekCursor::removeCursor(const TeamID& teamID) {
+	auto mapperIter = teamIDCursorMapper.find(teamID);
+	if (mapperIter == teamIDCursorMapper.end()) {
+		return nullptr;
+	}
+
+	// The iterator to the cursor in CursorContainer
+	auto iter = mapperIter->second;
+
+	// Remove from mapper
+	teamIDCursorMapper.erase(mapperIter);
+
+	// Remove from heap, there is no simple way of removing a specific item from a std::priority_queue, so we
+	// re-construct it.
+	std::vector<IndexedCursor> itemsInHeap;
+	while (!cursorHeap.empty()) {
+		itemsInHeap.emplace_back(cursorHeap.top());
+		cursorHeap.pop();
+	}
+	for (const auto& item : itemsInHeap) {
+		if (item.pCursorPtr == iter) {
+			continue;
+		}
+		cursorHeap.emplace(item);
+	}
+
+	std::unique_ptr<PeekCursorBase> pCursor(std::move(*iter));
+	cursorPtrs.erase(iter);
+
+	return pCursor;
+}
+
+Future<bool> MergedServerTeamPeekCursor::remoteMoreAvailableImpl() {
+	if (cursorPtrs.empty()) {
+		// No cursors are currently active
+		return false;
+	}
+
+	return peekRemoteForMergedServerTeamCursor({ &cursorPtrs, &cursorHeap, &teamIDCursorMapper });
+}
+
+std::vector<TeamID> MergedServerTeamPeekCursor::getCursorTeamIDs() {
+	std::vector<TeamID> result;
+	result.reserve(teamIDCursorMapper.size());
+	for (const auto& [teamID, _] : teamIDCursorMapper) {
+		result.push_back(teamID);
+	}
+	return result;
+}
+
+// Moves the cursor so it locates to the given version/subsequence. If the version/subsequence does not exist, moves the
+// cursor to the closest next mutation. If the version/subsequence is earlier than the current version/subsequence the
+// cursor is located, then the code will do nothing.
+ACTOR Future<Void> advanceTo(PeekCursorBase* cursor, Version version, Subsequence subsequence) {
+	state PeekCursorBase::iterator iter = cursor->begin();
+
+	loop {
+		while (iter != cursor->end()) {
+			// Is iter already past the version?
+			if (iter->version > version) {
+				return Void();
+			}
+			// Is iter current at the given version
+			if (iter->version == version) {
+				while (iter != cursor->end() && iter->subsequence < subsequence)
+					++iter;
```

Comment:
what if `iter->version > version`? It seems the code will just continue if the subsequence is smaller, which can be wrong. It's probably better to compare the pair `{iter->version, iter->subsequence}` is smaller than `{version, subsequence}` pair here.

Reply:
Indeed. Need to check the iter->version in the while loop too.